src/caffeine/collective_subroutines/co_reduce_s.f90

```diff
@@ -1,9 +1,227 @@
-submodule(collective_subroutines_m)  co_reduce_s
+! Copyright (c), The Regents of the University of California
+! Terms of use are as specified in LICENSE.txt
+submodule(collective_subroutines_m) co_reduce_s
+  use iso_c_binding, only : &
+    c_int64_t, c_ptr, c_size_t, c_loc, c_double, c_null_ptr, c_funptr, c_funloc, c_associated, c_f_pointer, c_int
+  use assert_m, only : assert
+  use intrinsic_array_m, only : intrinsic_array_t
+  use utilities_m, only : get_c_ptr, optional_value
+  implicit none
+
+  procedure(c_int32_t_operation), pointer :: c_int32_t_op_ptr => null()
+  procedure(c_float_operation), pointer :: c_float_op_ptr => null()
+  procedure(c_char_operation), pointer :: c_char_op_ptr => null()
+  procedure(c_bool_operation), pointer :: c_bool_op_ptr => null()
 
 contains
+ 
+  module procedure caf_co_reduce_c_char
+
+    interface
+
+      subroutine c_co_reduce_char(c_loc_a, Nelem, c_loc_stat, result_image, Coll_ReduceSub_c_char, client_data) bind(C)
+        import c_ptr, c_size_t, c_funptr, c_int
+        implicit none
+        type(c_ptr), value :: c_loc_a, c_loc_stat, Coll_ReduceSub_c_char, client_data
+        integer(c_size_t), value :: Nelem
+        integer(c_int), value :: result_image
+      end subroutine
+
+    end interface
+
+    type(c_ptr) stat_ptr
+
+    call assert(associated(operation), "caf_co_reduce_c_char: operation associated")
+
+    c_char_op_ptr => operation
+
+    stat_ptr = get_c_ptr(stat)
+
+    select rank(a)
+      rank(0)
+        block
+          integer(c_size_t), target :: len_a
+          len_a = int(len(a), c_size_t)
```

Comment:
IIUC, the requirement for A to have "the same shape, type, and type parameter values" means the character strings are guaranteed to be the same length on all images, right?